rfcs/006-package-management.md

```diff
@@ -0,0 +1,425 @@
+`github/jneem/foo` nickel package.
+
+Packages will be updated by pull requests to the index. There can be some tooling
+in the nickel CLI to automate the creation of these pull requests. If the volume
+of updates becomes high enough, we can automate the merging.
+
+### Question: should we store a content hash too?
+
+We're storing a git tree hash in the index, but if we ever want to store package
+contents in the future, maybe we should also store a hash of the git tree
+contents? This would allow verification of package tarballs, without needing the
+whole git repo.
+
+## Integration point: package maps
+
+Other tools (like build systems) may need to (1) consume the dependency resolution
+that we produce, or (2) plug in their own dependencies to the nickel interpreter.
+For (1), they can consume our lock-file, which will be in JSON. This means its
+format needs to be stable.
+
+For (2), we divide the package management implementation into two parts: the `plate`
+command has all the logic for consuming the manifest, fetching packages, and so on.
+The other part is to teach the nickel interpreter about "package maps," which is
+just a map associating a filesystem path to each pair
+`(package filesystem path, package-local name)`. When the nickel interpreter
+is running a file that came from the package at path `/home/jneem/foo`, and that
+file contains `import bar`, the nickel interpreter looks up `(/home/jneem/foo, bar)`
+in the package map to figure out where to import from.
+
+After `plate` fetches dependencies, it provides the interpreter with the correct
+package map to find them. Other tools with their own dependency-fetching methods
+can invoke the interpreter with a custom package map to makes those packages
+available.
+
+## CLI support
+
+We'll need some CLI commands for handling common package-management tasks. The
+current prototype has
+
+- a `plate package generate-lockfile` command that updates the lock-file
+- a `plate package debug-resolution` command that prints the full recursive
+  dependency tree
+
+We probably also want
+
+- a command for adding a new dependency to the manifest (checking if it exists,
+  and picking the most recent version)
+- a command for downloading the dependency tree (for use in build systems that
+  expect different "fetch" and "build" phases)
+- a command that checks for new dependency versions and updates the manifest
+- a command for automating the pull request necessary for updating your package
+
+Anything else?
+
+## Where do the packages live?
+
+Index and git dependencies need to be downloaded before they are used. We will
+use a single global cache directory for all dependencies. It will use the
+[`directories`](https://docs.rs/directories/latest/directories/struct.ProjectDirs.html#method.cache_dir)
+crate to deduce a platform-appropriate location (`~/.cache/nickel-lang/` on
+linux, unless `$XDG_CACHE_HOME` is set). Within this directory, git checkouts
```

Comment:
Do we need some garbage-collection/cache-eviction policy?

Reply:
Maybe, but it might also be ok to punt on it. The local cache doesn't have an index, so it's safe to just delete things manually.